File: ios/Storehop/Data/Demo/DemoDataSeeder.swift
import Foundation
import GRDB

#if DEBUG

/// DEBUG / marketing-only helper that fills the app with a curated, realistic
/// dataset so landing-page screenshots and the walkthrough tour look complete.
///
/// A fresh install ships with empty shelves — `DatabaseSeeder` seeds only
/// stores/categories/aisles, never items or purchase history — so this builds
/// everything on top via the public repositories (which keep the xref /
/// store-category-order / ownership invariants correct). Purchase history is
/// the one low-level exception: it's inserted straight through
/// `PurchaseRecordDao` with backdated `purchasedAt` so the Statistics screen
/// shows weeks of trend, because the repository purchase path can only stamp
/// "now".
///
/// Port of Android's `com.storehop.app.data.demo.DemoDataSeeder` — same
/// stores, categories, items, flags, and backdated purchase cadence, so the
/// two platforms produce visually matching marketing shots.
///
/// Reused by two consumers: the debug-only Settings loader
/// (`DemoDataDebugSection`) and the `DesignSystemTourTest` screenshot tour
/// (via the `-E2ESeedDemoData` launch argument). The whole file is compiled
/// out of release builds by the surrounding `#if DEBUG`, mirroring the way
/// R8 strips the Kotlin original.
///
/// Items intentionally carry no `imageUrl` — the UI renders its coloured
/// initial-avatar, which keeps the demo free of copyrighted brand imagery.
struct DemoDataSeeder: Sendable {
    /// Reader used only by `clear()` to snapshot the alive rows to remove.
    /// All content writes go through the repositories below.
    let writer: any DatabaseWriter
    let storeRepository: StoreRepository
    let categoryRepository: CategoryRepository
    let itemRepository: ItemRepository
    let storeCategoryOrderRepository: StoreCategoryOrderRepository
    let purchaseRecordDao: PurchaseRecordDao
    let ids: any IdGenerator
    let clock: any Clock
    let session: any UserSessionProvider
    let householdSession: any HouseholdSessionProvider

    // MARK: - Curated dataset

    /// One curated store.
    private struct StoreSpec {
        let key: String
        let name: String
        let colorArgb: Int64
        var oneOff: Bool = false
    }

    /// One curated item. `stores`/`category` reference the keys below.
    private struct ItemSpec {
        let name: String
        let category: String
        let stores: [String]
        var brand: String? = nil
        var staple: Bool = false
        var priority: Bool = false
        var buyToday: Bool = false
        /// Mark purchased "now" so it renders checked-off in the store list.
        var purchased: Bool = false
    }

    private let stores: [StoreSpec] = [
        StoreSpec(key: "continente", name: "Continente", colorArgb: 0xFF1565C0),
        StoreSpec(key: "pingo", name: "Pingo Doce", colorArgb: 0xFF2E7D32),
        StoreSpec(key: "lidl", name: "Lidl", colorArgb: 0xFFF9A825),
        StoreSpec(key: "aldi", name: "Aldi", colorArgb: 0xFF00838F),
        StoreSpec(key: "farmacia", name: "Farmácia", colorArgb: 0xFFC62828, oneOff: true),
    ]

    // Ordered — the list order becomes each store's default aisle order.
    private let categories: [String] = [
        "Produce", "Bakery", "Dairy", "Meat & Fish",
        "Frozen", "Pantry", "Household", "Toiletries",
    ]

    private let items: [ItemSpec] = [
        // Produce
        ItemSpec(name: "Bananas", category: "Produce", stores: ["continente", "pingo"]),
        ItemSpec(name: "Baby Spinach", category: "Produce", stores: ["continente"], brand: "Florette"),
        ItemSpec(name: "Avocados", category: "Produce", stores: ["pingo", "lidl"]),
        ItemSpec(name: "Tomatoes", category: "Produce", stores: ["continente"], purchased: true),
        // Bakery
        ItemSpec(name: "Sourdough Bread", category: "Bakery", stores: ["continente", "pingo"], staple: true),
        ItemSpec(name: "Croissants", category: "Bakery", stores: ["lidl"]),
        // Dairy
        ItemSpec(name: "Whole Milk", category: "Dairy", stores: ["continente", "pingo", "aldi"], brand: "Mimosa", staple: true),
        ItemSpec(name: "Greek Yogurt", category: "Dairy", stores: ["continente"], brand: "Fage"),
        ItemSpec(name: "Butter", category: "Dairy", stores: ["pingo"], staple: true),
        ItemSpec(name: "Eggs", category: "Dairy", stores: ["continente", "aldi"], staple: true, purchased: true),
        // Meat & Fish
        ItemSpec(name: "Chicken Breasts", category: "Meat & Fish", stores: ["continente", "pingo"]),
        ItemSpec(name: "Salmon Fillets", category: "Meat & Fish", stores: ["continente"]),
        ItemSpec(name: "Ground Beef", category: "Meat & Fish", stores: ["aldi"]),
        // Frozen
        ItemSpec(name: "Chicken Wings - Frozen", category: "Frozen", stores: ["lidl", "aldi"]),
        ItemSpec(name: "Garden Peas", category: "Frozen", stores: ["lidl"]),
        ItemSpec(name: "Ice Cream", category: "Frozen", stores: ["continente"], brand: "Olá"),
        // Pantry
        ItemSpec(name: "Coffee", category: "Pantry", stores: ["continente", "pingo"], brand: "Delta", staple: true),
        ItemSpec(name: "Olive Oil", category: "Pantry", stores: ["continente"], brand: "Gallo"),
        ItemSpec(name: "Spaghetti", category: "Pantry", stores: ["lidl", "aldi"], purchased: true),
        ItemSpec(name: "Basmati Rice", category: "Pantry", stores: ["continente"]),
        ItemSpec(name: "Sugar", category: "Pantry", stores: ["pingo", "aldi"]),
        // Household
        ItemSpec(name: "Toilet Paper", category: "Household", stores: ["continente", "pingo"], brand: "Renova", buyToday: true),
        ItemSpec(name: "Dish Soap", category: "Household", stores: ["continente"], brand: "Fairy"),
        ItemSpec(name: "Trash Bags", category: "Household", stores: ["lidl"]),
        ItemSpec(name: "Laundry Detergent", category: "Household", stores: ["aldi"], brand: "Skip", priority: true),
        // Toiletries
        ItemSpec(name: "Toothpaste", category: "Toiletries", stores: ["continente"], brand: "Colgate"),
        ItemSpec(name: "Shampoo", category: "Toiletries", stores: ["pingo"]),
        ItemSpec(name: "Razors", category: "Toiletries", stores: ["continente"], brand: "Gillette"),
        ItemSpec(name: "Ibuprofen", category: "Toiletries", stores: ["farmacia"], brand: "Brufen", priority: true, buyToday: true),
    ]

    // MARK: - Seed

    /// Wipe any existing alive content, then build the curated demo dataset in
    /// the current session scope. Safe to run on an empty DB (screenshot tour)
    /// or on the seeded debug app (archives the shipped stores/categories first
    /// so only the demo set shows).
    func seed() async throws {
        try await clear()

        // Categories first — the item + aisle-order writes reference them.
        // Insert sequentially so `displayOrder` lands in list order (each
        // `addCategory` appends at MAX(displayOrder)+1).
        var categoryIds: [String: String] = [:]
        for name in categories {
            categoryIds[name] = try await categoryRepository.addCategory(name: name, icon: nil)
        }

        // Stores next, sequentially so `displayOrder` follows the array order.
        var storeIds: [String: String] = [:]
        for spec in stores {
            storeIds[spec.key] = try await storeRepository.addStore(
                name: spec.name,
                colorArgb: spec.colorArgb,
                isOneOff: spec.oneOff
            )
        }

        // Give each store a sensible aisle order (the category list order).
        let orderedCategoryIds = categories.compactMap { categoryIds[$0] }
        for storeId in storeIds.values {
            try await storeCategoryOrderRepository.reorderCategoriesForStore(
                storeId: storeId,
                orderedCategoryIds: orderedCategoryIds
            )
        }

        var itemIdByName: [String: String] = [:]
        for spec in items {
            let itemId = try await itemRepository.addItem(
                name: spec.name,
                categoryId: categoryIds[spec.category],
                storeIds: Set(spec.stores.compactMap { storeIds[$0] }),
                quantity: nil,
                notes: nil,
                brand: spec.brand,
                imageUrl: nil,
                isStaple: spec.staple,
                isPriority: spec.priority,
                isBuyToday: spec.buyToday
            )
            itemIdByName[spec.name] = itemId
            if spec.purchased {
                // Renders struck-through in the store list (session-purchased
                // window) and contributes a "now" purchase to the stats.
                if let storeId = spec.stores.compactMap({ storeIds[$0] }).first {
                    try await itemRepository.markPurchasedAtStore(itemId: itemId, storeId: storeId)
                }
            }
        }

        try await seedPurchaseHistory(itemIdByName: itemIdByName, storeIdByKey: storeIds)
    }

    /// Backdated purchase history so Statistics shows weeks of trend, top items,
    /// per-store and per-category bars. Deterministic (no RNG) so screenshots are
    /// reproducible. Skips silently when there's no active session.
    ///
    /// Anchored on `clock.nowMs()` — the same instant `StatisticsViewModel`
    /// uses to compute its trailing windows — so the backdated records always
    /// land inside the trend/last-30/last-7 ranges regardless of whether the
    /// clock is the live `SystemClock` or the E2E `FixedClock`.
    private func seedPurchaseHistory(
        itemIdByName: [String: String],
        storeIdByKey: [String: String]
    ) async throws {
        guard let userId = await session.currentUserId else { return }
        let householdId = (await householdSession.currentHouseholdId) ?? userId
        let now = clock.nowMs()
        let day: Int64 = 24 * 60 * 60 * 1000

        // Items that get bought often (drive "top items" + the trend line),
        // paired with the store they're usually bought at.
        let recurring: [(item: String, store: String)] = [
            ("Whole Milk", "continente"),
            ("Eggs", "continente"),
            ("Sourdough Bread", "pingo"),
            ("Bananas", "continente"),
            ("Coffee", "pingo"),
            ("Butter", "pingo"),
            ("Chicken Breasts", "continente"),
            ("Greek Yogurt", "continente"),
            ("Spaghetti", "lidl"),
            ("Ice Cream", "continente"),
        ]

        // Build the full record set first, then commit in a single
        // transaction (iOS idiom for bulk inserts; the Kotlin original loops
        // `dao.insert` one row at a time).
        var records: [PurchaseRecord] = []
        var count = 0
        // Spread ~8 weeks of shopping trips. Each "trip" is a day a few times
        // a week; on each trip a rotating handful of the recurring items is
        // bought.
        for dayAgo in 1...56 {
            // Shop roughly Mon/Thu/Sat — pick 3 cadence days per week.
            guard [0, 3, 5].contains(dayAgo % 7) else { continue }
            let tripItems = recurring.enumerated().filter { (idx, _) in (idx + dayAgo) % 3 == 0 }
            for (_, entry) in tripItems {
                guard let itemId = itemIdByName[entry.item],
                      let storeId = storeIdByKey[entry.store] else { continue }
                let purchasedAt = now - Int64(dayAgo) * day - Int64(count % 6) * 60 * 60 * 1000
                records.append(
                    PurchaseRecord(
                        id: ids.newId(),
                        itemId: itemId,
                        storeId: storeId,
                        purchasedAt: purchasedAt,
                        userId: userId,
                        createdAt: now,
                        updatedAt: now,
                        deletedAt: nil,
                        pendingSync: true,
                        householdId: householdId
                    )
                )
                count += 1
            }
        }

        guard !records.isEmpty else { return }
        try await writer.write { db in
            for record in records {
                try PurchaseRecordDao.insert(record, on: db)
            }
        }
    }

    // MARK: - Clear

    /// Remove demo (and any other) alive content so a re-seed starts clean:
    /// soft-delete every alive item and archive every alive store + category.
    /// A fresh install still restores the shipped seed; this is debug-only.
    func clear() async throws {
        // The repository read/observe surface is scoped by householdId; the
        // repositories forward the value they receive as `userId:`. Snapshot
        // the alive ids directly (a one-shot read) rather than pulling the
        // first element off each `AsyncValueObservation`, which has no clean
        // "first value" accessor.
        let householdId = try await householdSession.requireHouseholdId()

        let itemIds = try await writer.read { db in
            try String.fetchAll(db, sql: """
                SELECT id FROM items
                WHERE householdId = ? AND deletedAt IS NULL
                """, arguments: [householdId])
        }
        for id in itemIds {
            try await itemRepository.softDelete(id: id)
        }

        let storeIds = try await writer.read { db in
            try String.fetchAll(db, sql: """
                SELECT id FROM stores
                WHERE householdId = ? AND deletedAt IS NULL AND isArchived = 0
                """, arguments: [householdId])
        }
        for id in storeIds {
            try await storeRepository.setArchived(id: id, archived: true)
        }

        let categoryIds = try await writer.read { db in
            try String.fetchAll(db, sql: """
                SELECT id FROM categories
                WHERE householdId = ? AND deletedAt IS NULL AND isArchived = 0
                """, arguments: [householdId])
        }
        for id in categoryIds {
            try await categoryRepository.setArchived(id: id, archived: true)
        }
    }
}

#endif

File: app/src/main/java/com/storehop/app/data/demo/DemoDataSeeder.kt
package com.storehop.app.data.demo

import com.storehop.app.data.dao.PurchaseRecordDao
import com.storehop.app.data.entity.PurchaseRecord
import com.storehop.app.data.repository.CategoryRepository
import com.storehop.app.data.repository.ItemRepository
import com.storehop.app.data.repository.StoreCategoryOrderRepository
import com.storehop.app.data.repository.StoreRepository
import com.storehop.app.data.util.HouseholdSessionProvider
import com.storehop.app.data.util.IdGenerator
import com.storehop.app.data.util.UserSessionProvider
import kotlinx.coroutines.flow.first
import java.time.Clock
import javax.inject.Inject
import javax.inject.Singleton

/**
 * DEBUG / marketing-only helper that fills the app with a curated, realistic
 * dataset so landing-page screenshots and the walkthrough video look complete.
 *
 * A fresh install ships with empty shelves — [com.storehop.app.data.db.DatabaseSeeder]
 * seeds only stores/categories/aisles, never items or purchase history — so
 * this builds everything on top via the public repositories (which keep the
 * xref / store-category-order / ownership invariants correct). Purchase history
 * is the one low-level exception: it's inserted straight through
 * [PurchaseRecordDao] with backdated `purchasedAt` so the Statistics screen
 * shows weeks of trend, because the repository purchase path can only stamp
 * "now".
 *
 * Reused by three consumers: the debug-only Settings loader, the
 * `ScreenshotTourTest`, and the `DemoFlowTest` recorder. It is only ever
 * invoked behind `BuildConfig.DEBUG` (Settings) or from instrumented tests, so
 * R8 strips it from the release build.
 *
 * Items intentionally carry no `imageUrl` — the UI renders its coloured
 * initial-avatar, which keeps the demo free of copyrighted brand imagery.
 */
@Singleton
class DemoDataSeeder @Inject constructor(
    private val storeRepository: StoreRepository,
    private val categoryRepository: CategoryRepository,
    private val itemRepository: ItemRepository,
    private val storeCategoryOrderRepository: StoreCategoryOrderRepository,
    private val purchaseRecordDao: PurchaseRecordDao,
    private val ids: IdGenerator,
    private val clock: Clock,
    private val session: UserSessionProvider,
    private val householdSession: HouseholdSessionProvider,
) {

    /** One curated store. */
    private data class StoreSpec(val key: String, val name: String, val colorArgb: Int, val oneOff: Boolean = false)

    /** One curated item. `stores`/`category` reference the keys below. */
    private data class ItemSpec(
        val name: String,
        val category: String,
        val stores: List<String>,
        val brand: String? = null,
        val staple: Boolean = false,
        val priority: Boolean = false,
        val buyToday: Boolean = false,
        /** Mark purchased "now" so it renders checked-off in the store list. */
        val purchased: Boolean = false,
    )

    private val stores = listOf(
        StoreSpec("continente", "Continente", 0xFF1565C0.toInt()),
        StoreSpec("pingo", "Pingo Doce", 0xFF2E7D32.toInt()),
        StoreSpec("lidl", "Lidl", 0xFFF9A825.toInt()),
        StoreSpec("aldi", "Aldi", 0xFF00838F.toInt()),
        StoreSpec("farmacia", "Farmácia", 0xFFC62828.toInt(), oneOff = true),
    )

    // Ordered — the list order becomes each store's default aisle order.
    private val categories = listOf(
        "Produce", "Bakery", "Dairy", "Meat & Fish",
        "Frozen", "Pantry", "Household", "Toiletries",
    )

    private val items = listOf(
        // Produce
        ItemSpec("Bananas", "Produce", listOf("continente", "pingo")),
        ItemSpec("Baby Spinach", "Produce", listOf("continente"), brand = "Florette"),
        ItemSpec("Avocados", "Produce", listOf("pingo", "lidl")),
        ItemSpec("Tomatoes", "Produce", listOf("continente"), purchased = true),
        // Bakery
        ItemSpec("Sourdough Bread", "Bakery", listOf("continente", "pingo"), staple = true),
        ItemSpec("Croissants", "Bakery", listOf("lidl")),
        // Dairy
        ItemSpec("Whole Milk", "Dairy", listOf("continente", "pingo", "aldi"), brand = "Mimosa", staple = true),
        ItemSpec("Greek Yogurt", "Dairy", listOf("continente"), brand = "Fage"),
        ItemSpec("Butter", "Dairy", listOf("pingo"), staple = true),
        ItemSpec("Eggs", "Dairy", listOf("continente", "aldi"), staple = true, purchased = true),
        // Meat & Fish
        ItemSpec("Chicken Breasts", "Meat & Fish", listOf("continente", "pingo")),
        ItemSpec("Salmon Fillets", "Meat & Fish", listOf("continente")),
        ItemSpec("Ground Beef", "Meat & Fish", listOf("aldi")),
        // Frozen
        ItemSpec("Chicken Wings - Frozen", "Frozen", listOf("lidl", "aldi")),
        ItemSpec("Garden Peas", "Frozen", listOf("lidl")),
        ItemSpec("Ice Cream", "Frozen", listOf("continente"), brand = "Olá"),
        // Pantry
        ItemSpec("Coffee", "Pantry", listOf("continente", "pingo"), brand = "Delta", staple = true),
        ItemSpec("Olive Oil", "Pantry", listOf("continente"), brand = "Gallo"),
        ItemSpec("Spaghetti", "Pantry", listOf("lidl", "aldi"), purchased = true),
        ItemSpec("Basmati Rice", "Pantry", listOf("continente")),
        ItemSpec("Sugar", "Pantry", listOf("pingo", "aldi")),
        // Household
        ItemSpec("Toilet Paper", "Household", listOf("continente", "pingo"), brand = "Renova", buyToday = true),
        ItemSpec("Dish Soap", "Household", listOf("continente"), brand = "Fairy"),
        ItemSpec("Trash Bags", "Household", listOf("lidl")),
        ItemSpec("Laundry Detergent", "Household", listOf("aldi"), brand = "Skip", priority = true),
        // Toiletries
        ItemSpec("Toothpaste", "Toiletries", listOf("continente"), brand = "Colgate"),
        ItemSpec("Shampoo", "Toiletries", listOf("pingo")),
        ItemSpec("Razors", "Toiletries", listOf("continente"), brand = "Gillette"),
        ItemSpec("Ibuprofen", "Toiletries", listOf("farmacia"), brand = "Brufen", priority = true, buyToday = true),
    )

    /**
     * Wipe any existing alive content, then build the curated demo dataset in
     * the current session scope. Safe to run on an empty DB (screenshot tests)
     * or on the seeded debug app (archives the shipped stores/categories first
     * so only the demo set shows).
     */
    suspend fun seed() {
        clear()

        val categoryIds = categories.associateWith { name ->
            categoryRepository.addCategory(name = name, icon = null)
        }
        val storeIds = stores.associate { spec ->
            spec.key to storeRepository.addStore(spec.name, spec.colorArgb, spec.oneOff)
        }

        // Give each store a sensible aisle order (the category list order).
        val orderedCategoryIds = categories.mapNotNull { categoryIds[it] }
        storeIds.values.forEach { storeId ->
            storeCategoryOrderRepository.reorderCategoriesForStore(storeId, orderedCategoryIds)
        }

        val itemIdByName = LinkedHashMap<String, String>()
        for (spec in items) {
            val itemId = itemRepository.addItem(
                name = spec.name,
                categoryId = categoryIds[spec.category],
                storeIds = spec.stores.mapNotNull { storeIds[it] }.toSet(),
                brand = spec.brand,
                isStaple = spec.staple,
                isPriority = spec.priority,
                isBuyToday = spec.buyToday,
            )
            itemIdByName[spec.name] = itemId
            if (spec.purchased) {
                // Renders struck-through in the store list (session-purchased
                // window) and contributes a "now" purchase to the stats.
                val storeId = spec.stores.firstNotNullOfOrNull { storeIds[it] }
                if (storeId != null) itemRepository.markPurchasedAtStore(itemId, storeId)
            }
        }

        seedPurchaseHistory(itemIdByName, storeIds)
    }

    /**
     * Backdated purchase history so Statistics shows weeks of trend, top items,
     * per-store and per-category bars. Deterministic (no RNG) so screenshots are
     * reproducible. Skips silently when there's no active session.
     */
    private suspend fun seedPurchaseHistory(
        itemIdByName: Map<String, String>,
        storeIdByKey: Map<String, String>,
    ) {
        val userId = session.currentUserId() ?: return
        val householdId = householdSession.currentHouseholdId() ?: userId
        val now = clock.millis()
        val day = 24L * 60 * 60 * 1000

        // Items that get bought often (drive "top items" + the trend line),
        // paired with the store they're usually bought at.
        val recurring = listOf(
            "Whole Milk" to "continente",
            "Eggs" to "continente",
            "Sourdough Bread" to "pingo",
            "Bananas" to "continente",
            "Coffee" to "pingo",
            "Butter" to "pingo",
            "Chicken Breasts" to "continente",
            "Greek Yogurt" to "continente",
            "Spaghetti" to "lidl",
            "Ice Cream" to "continente",
        )

        // Spread ~12 weeks of shopping trips with an organic (but deterministic,
        // so screenshots are reproducible) cadence: not every day is a trip, and
        // trip sizes vary — so the "daily activity" chart reads like a real user
        // instead of a regular sawtooth.
        var records = 0
        for (dayAgo in 1..84) {
            val roll = hash(dayAgo)
            // Shop ~4 days in 7; skip the rest.
            if (roll % 7 < 3) continue
            val tripSize = 2 + (hash(dayAgo * 31 + 7) % 6) // 2..7 items this trip
            val chosen = LinkedHashSet<Int>()
            var k = 0
            while (chosen.size < tripSize && k < tripSize * 3) {
                chosen.add(hash(dayAgo * 17 + k) % recurring.size)
                k++
            }
            for ((slot, idx) in chosen.withIndex()) {
                val (itemName, storeKey) = recurring[idx]
                val itemId = itemIdByName[itemName] ?: continue
                val storeId = storeIdByKey[storeKey] ?: continue
                // Stagger within the day so records don't share a timestamp.
                val purchasedAt = now - dayAgo * day + slot * 37L * 60 * 1000
                purchaseRecordDao.insert(
                    PurchaseRecord(
                        id = ids.newId(),
                        itemId = itemId,
                        storeId = storeId,
                        purchasedAt = purchasedAt,
                        userId = userId,
                        createdAt = now,
                        updatedAt = now,
                        deletedAt = null,
                        pendingSync = true,
                        householdId = householdId,
                    ),
                )
                records++
            }
        }
    }

    /** Deterministic non-negative pseudo-random — reproducible history layout. */
    private fun hash(seed: Int): Int {
        var x = seed * 1103515245 + 12345
        x = x xor (x ushr 16)
        x *= -2048144789
        x = x xor (x ushr 13)
        return x and 0x7fffffff
    }

    /**
     * Remove demo (and any other) alive content so a re-seed starts clean:
     * soft-delete every alive item and archive every alive store + category.
     * A fresh install still restores the shipped seed; this is debug-only.
     */
    suspend fun clear() {
        itemRepository.observeAll().first().forEach { itemRepository.softDelete(it.item.id) }
        storeRepository.observeAll(includeArchived = false).first().forEach {
            storeRepository.setArchived(it.id, true)
        }
        categoryRepository.observeAll(includeArchived = false).first().forEach {
            categoryRepository.setArchived(it.id, true)
        }
    }
}
